File: runtimes/python/facet.py
"""Facet Architecture - Python server runtime.

The Python analogue of the Go ``fa`` package: it turns the compiler's neutral
output (``manifest.json`` + ``render.json``, from ``fct build`` with
``target = "python"``) into a live app - SSE transport, HMAC-signed event push,
the render-IR interpreter, and the ``/events`` router. The wire format, signing
layout, and client runtime are shared with every other target (see
docs/BACKENDS.md, fa/wire.go, fa/event.go).

Zero dependencies: standard library only.
"""

import hashlib
import hmac
import html
import json
import os
import queue
import re
import secrets
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

import native

WIRE_VERSION = "1"


def dumps(obj):
    """Compact JSON, matching the wire bytes the other runtimes emit (no spaces)."""
    return json.dumps(obj, separators=(",", ":"))


# -- render IR -> block tree -------------------------------------------------
# The flat op stream is parsed once into a nested tree for a clean recursive walk.

def build_tree(ops):
    pos = [0]

    def block():
        nodes = []
        while pos[0] < len(ops):
            op = ops[pos[0]]
            if op["op"] in ("end", "else"):
                return nodes
            pos[0] += 1
            kind = op["op"]
            if kind == "text":
                nodes.append({"t": "text", "v": op.get("v", "")})
            elif kind == "expr":
                nodes.append({"t": "expr", "x": op["x"]})
            elif kind == "child":
                nodes.append({"t": "child", "name": op["name"], "props": op.get("props", [])})
            elif kind == "if":
                then_b = block()
                els_b = []
                if pos[0] < len(ops) and ops[pos[0]]["op"] == "else":
                    pos[0] += 1
                    els_b = block()
                if pos[0] < len(ops) and ops[pos[0]]["op"] == "end":
                    pos[0] += 1
                nodes.append({"t": "if", "x": op["x"], "then": then_b, "els": els_b})
            elif kind == "for":
                body = block()
                if pos[0] < len(ops) and ops[pos[0]]["op"] == "end":
                    pos[0] += 1
                nodes.append({"t": "for", "var": op["var"], "x": op["x"], "body": body})
        return nodes

    return block()


# -- expression evaluator (neutral irExpr AST) -------------------------------

def eval_expr(x, scope):
    k = x["k"]
    if k == "num":
        n = x.get("n", "0")
        return float(n) if "." in n else int(n)
    if k == "str":
        return x.get("s", "")
    if k == "bool":
        return bool(x.get("b", False))
    if k == "path":
        return eval_path(x, scope)
    if k == "call":
        fn = eval_path(x["recv"], scope)
        args = [eval_expr(a, scope) for a in x.get("args", [])]
        return fn(*args) if callable(fn) else None
    if k == "unary":
        if x["op"] == "!":
            return not truthy(eval_expr(x["x"], scope))
        return -_num(eval_expr(x["x"], scope))
    if k == "bin":
        return eval_bin(x["op"], eval_expr(x["l"], scope), eval_expr(x["r"], scope))
    return None


def eval_path(x, scope):
    segs = x.get("segs", [])
    src = scope["locals"] if x.get("local") else scope["data"]
    cur = _get(src, segs[0])
    for seg in segs[1:]:
        if cur is None:
            return None
        cur = _get(cur, seg)
    return cur


def _get(obj, key):
    if isinstance(obj, dict):
        return obj.get(key)
    return getattr(obj, key, None)


def eval_bin(op, a, b):
    if op == "==":
        return a == b
    if op == "!=":
        return a != b
    if op == "<":
        return _num(a) < _num(b)
    if op == "<=":
        return _num(a) <= _num(b)
    if op == ">":
        return _num(a) > _num(b)
    if op == ">=":
        return _num(a) >= _num(b)
    if op == "&&":
        return truthy(a) and truthy(b)
    if op == "||":
        return a if truthy(a) else b
    if op == "+":
        if isinstance(a, str) or isinstance(b, str):
            return _fmt(a) + _fmt(b)
        return _num(a) + _num(b)
    if op == "-":
        return _num(a) - _num(b)
    if op == "*":
        return _num(a) * _num(b)
    if op == "/":
        return _num(a) / _num(b)
    if op == "%":
        return _num(a) % _num(b)
    return None


def _num(v):
    if isinstance(v, bool):
        return 1 if v else 0
    if isinstance(v, (int, float)):
        return v
    try:
        return float(v)
    except (TypeError, ValueError):
        return 0


def truthy(v):
    return bool(v)


def _fmt(v):
    if v is None:
        return ""
    if isinstance(v, bool):
        return "true" if v else "false"
    return str(v)


def html_escape(s):
    return html.escape(_fmt(s), quote=True)


# -- renderer ----------------------------------------------------------------

def render_nodes(nodes, scope, app):
    out = []
    for n in nodes:
        t = n["t"]
        if t == "text":
            out.append(n["v"])
        elif t == "expr":
            out.append(html_escape(eval_expr(n["x"], scope)))
        elif t == "if":
            branch = n["then"] if truthy(eval_expr(n["x"], scope)) else n["els"]
            out.append(render_nodes(branch, scope, app))
        elif t == "for":
            items = eval_expr(n["x"], scope) or []
            for item in items:
                locals_ = dict(scope["locals"])
                locals_[n["var"]] = item
                out.append(render_nodes(n["body"], {"data": scope["data"], "locals": locals_}, app))
        elif t == "child":
            child_data = {}
            for p in n["props"]:
                child_data[p["name"]] = eval_expr(p["x"], scope) if "x" in p and p["x"] else p.get("lit", "")
            out.append(app.render_facet(n["name"], child_data)["html"])
    return "".join(out)


_TAG_RE = re.compile(r"<([a-zA-Z][\w-]*)")
_HOLE_RE = re.compile(r"\{([^}]+)\}")


def resolve_facet_id(pattern, data):
    def repl(m):
        segs = m.group(1).split(".")
        cur = _get(data, segs[0])
        for seg in segs[1:]:
            if cur is None:
                break
            cur = _get(cur, seg)
        return _fmt(cur)

    return _HOLE_RE.sub(repl, pattern)


def inject_facet_id(html_str, fid):
    return _TAG_RE.sub(lambda m: m.group(0) + ' data-facet-id="%s"' % fid, html_str, count=1)


# -- signing (mirrors fa/event.go sign(): op \0 facet_id \0 fragment) --------

def sign_event(key_hex, ev):
    if not key_hex:
        return ev
    mac = hmac.new(bytes.fromhex(key_hex), digestmod=hashlib.sha256)
    mac.update(ev["op"].encode())
    mac.update(b"\x00")
    mac.update(ev["facet_id"].encode())
    mac.update(b"\x00")
    mac.update(ev["fragment"].encode())
    ev["hmac"] = mac.hexdigest()
    return ev


# -- App ---------------------------------------------------------------------

class App:
    def __init__(self, gen_dir="generated", fa_key=None, title="FA", runtime_js=None):
        self.gen_dir = gen_dir
        self.key_hex = fa_key if fa_key is not None else os.environ.get("FA_SIGNING_KEY", "")
        self.title = title
        here = os.path.dirname(os.path.abspath(__file__))
        self.runtime_js = runtime_js or os.path.join(here, "..", "..", "runtime", "fa-runtime.js")
        with open(os.path.join(gen_dir, "manifest.json")) as f:
            self.manifest = json.load(f)
        with open(os.path.join(gen_dir, "render.json")) as f:
            self.ir = json.load(f)
        self.trees = {}
        self.facets = {}
        for fac in self.ir["facets"]:
            self.facets[fac["name"]] = fac
            self.trees[fac["name"]] = build_tree(fac["render"])
        self.handlers = {}
        self._root = None
        self.conns = {}  # conn id -> queue.Queue
        self._lock = threading.Lock()

    def root(self, name, data_fn):
        self._root = (name, data_fn)
        return self

    def on(self, event_type, fn):
        self.handlers[event_type] = fn
        return self

    def render_facet(self, name, data):
        tree = self.trees.get(name)
        if tree is None:
            raise KeyError("unknown facet " + name)
        fid = resolve_facet_id(self.facets[name]["facet_id"], data)
        body = render_nodes(tree, {"data": data, "locals": {}}, self)
        return {"facet_id": fid, "html": inject_facet_id(body, fid)}

    def tree_fragment(self, html_str):
        """The neutral ViewNode tree of a rendered facet, as a JSON string (the
        native fragment) — mirror of Go's ParseView."""
        return dumps(native.node_to_json(native.parse_view(html_str)))

    def push_rerender(self, event_type, conn, data):
        c = self.conns.get(conn)
        if c is None:
            return
        for fac in self.ir["facets"]:
            for w in fac.get("when", []):
                if event_type not in w["events"]:
                    continue
                for mu in w["mutations"]:
                    target = mu.get("target") or fac["name"]
                    r = self.render_facet(target, data)
                    op = "replace" if mu["op"] == "replace_all" else mu["op"]
                    # Native connection → ViewNode-tree fragment; web → HTML.
                    fragment = self.tree_fragment(r["html"]) if c["native"] else r["html"]
                    ev = sign_event(self.key_hex, {"op": op, "facet_id": r["facet_id"], "fragment": fragment})
                    c["q"].put("data: " + dumps(ev) + "\n\n")

    def listen(self, addr=None):
        addr = addr or os.environ.get("FA_ADDR", "localhost:7373")
        host, _, port = addr.partition(":")
        app = self

        class Handler(BaseHTTPRequestHandler):
            def log_message(self, *a):
                pass

            def do_GET(self):
                p = self.path.split("?")[0]
                if p == "/":
                    return app._serve_shell(self)
                if p == "/sse":
                    return app._serve_sse(self)
                if p == "/manifest.json":
                    return app._json(self, app.manifest)
                if p == "/render.json":
                    return app._json(self, app.ir)
                if p == "/fa-runtime.js":
                    with open(app.runtime_js, "rb") as f:
                        body = f.read()
                    self.send_response(200)
                    self.send_header("Content-Type", "text/javascript")
                    self.send_header("Content-Length", str(len(body)))
                    self.end_headers()
                    self.wfile.write(body)
                    return
                self.send_response(404)
                self.end_headers()

            def do_POST(self):
                if self.path.split("?")[0] == "/events":
                    return app._handle_events(self)
                self.send_response(404)
                self.end_headers()

        server = ThreadingHTTPServer((host, int(port)), Handler)
        print("fa(python): listening on http://%s:%s" % (host, port))
        server.serve_forever()

    # -- handlers ------------------------------------------------------------
    def _serve_shell(self, h):
        # Native client (FA-Native: 1) → ScreenResponse {title, tree}; browser → HTML.
        if h.headers.get("FA-Native") == "1":
            tree = {"kind": "box"}
            if self._root:
                name, data_fn = self._root
                html_str = self.render_facet(name, data_fn({}))["html"]
                tree = native.node_to_json(native.parse_view(html_str))
            return self._json(h, {"title": self.title, "tree": tree})
        body = ""
        if self._root:
            name, data_fn = self._root
            body = self.render_facet(name, data_fn({}))["html"]
        page = ('<!doctype html><html><head><meta charset="utf-8">'
                '<meta name="fa-key" content="%s"><title>%s</title></head><body>%s'
                '<script src="/fa-runtime.js"></script></body></html>'
                % (self.key_hex, html_escape(self.title), body))
        data = page.encode()
        h.send_response(200)
        h.send_header("Content-Type", "text/html; charset=utf-8")
        h.send_header("Content-Length", str(len(data)))
        h.end_headers()
        h.wfile.write(data)

    def _serve_sse(self, h):
        v = "1"
        if "?" in h.path:
            for kv in h.path.split("?")[1].split("&"):
                if kv.startswith("v="):
                    v = kv[2:]
        if v != WIRE_VERSION:
            h.send_response(426)
            h.end_headers()
            return
        h.send_response(200)
        h.send_header("Content-Type", "text/event-stream")
        h.send_header("Cache-Control", "no-cache")
        h.send_header("Connection", "keep-alive")
        h.send_header("X-Accel-Buffering", "no")
        h.end_headers()
        cid = secrets.token_hex(16)
        q = queue.Queue()
        with self._lock:
            self.conns[cid] = {"q": q, "native": h.headers.get("FA-Native") == "1"}
        hello = dumps({"op": "_conn", "conn": cid, "key": self.key_hex, "v": WIRE_VERSION})
        try:
            h.wfile.write(("data: %s\n\n" % hello).encode())
            h.wfile.flush()
            while True:
                try:
                    msg = q.get(timeout=25)
                except queue.Empty:
                    msg = ": keepalive\n\n"
                h.wfile.write(msg.encode())
                h.wfile.flush()
        except (BrokenPipeError, ConnectionResetError):
            pass
        finally:
            with self._lock:
                self.conns.pop(cid, None)

    def _handle_events(self, h):
        length = int(h.headers.get("Content-Length", 0))
        raw = h.rfile.read(length) if length else b"{}"
        h.send_response(204)
        h.end_headers()
        try:
            body = json.loads(raw or b"{}")
        except ValueError:
            return
        fn = self.handlers.get(body.get("type"))
        if not fn:
            return
        ctx = {"type": body.get("type"), "payload": body.get("payload") or {}, "conn": body.get("conn")}
        data = fn(ctx)
        if data is None:
            return
        self.push_rerender(ctx["type"], ctx["conn"], data)

    def _json(self, h, obj):
        data = dumps(obj).encode()
        h.send_response(200)
        h.send_header("Content-Type", "application/json")
        h.send_header("Content-Length", str(len(data)))
        h.end_headers()
        h.wfile.write(data)

File: runtimes/node/facet.js
'use strict';
// Facet Architecture — Node.js server runtime.
//
// The Node analogue of the Go `fa/` package: it turns the compiler's neutral
// output (manifest.json + render.json, from `fct build` with target = "node")
// into a live app. It owns the SSE transport, HMAC-signed event push, the
// render-IR interpreter, and the /events router — app code never hand-writes
// streaming or signing. The wire format, signing layout, and client runtime are
// shared with every other target (see docs/BACKENDS.md, fa/wire.go, fa/event.go).
//
// Zero dependencies: only Node built-ins (http, crypto, fs).

const http = require('http');
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');
const native = require('./native');

const WIRE_VERSION = '1';

// ── render IR → block tree ──────────────────────────────────────────────────
// The flat op stream (text/expr/if/else/end/for/child) is parsed once into a
// nested tree so rendering is a clean recursive walk.

function buildTree(ops) {
  let pos = 0;
  function block() {
    const nodes = [];
    while (pos < ops.length) {
      const op = ops[pos];
      if (op.op === 'end' || op.op === 'else') return nodes;
      pos++;
      switch (op.op) {
        case 'text': nodes.push({ t: 'text', v: op.v }); break;
        case 'expr': nodes.push({ t: 'expr', x: op.x }); break;
        case 'child': nodes.push({ t: 'child', name: op.name, props: op.props || [] }); break;
        case 'if': {
          const thenB = block();
          let elseB = [];
          if (ops[pos] && ops[pos].op === 'else') { pos++; elseB = block(); }
          if (ops[pos] && ops[pos].op === 'end') pos++;
          nodes.push({ t: 'if', x: op.x, then: thenB, els: elseB });
          break;
        }
        case 'for': {
          const body = block();
          if (ops[pos] && ops[pos].op === 'end') pos++;
          nodes.push({ t: 'for', var: op.var, x: op.x, body: body });
          break;
        }
      }
    }
    return nodes;
  }
  return block();
}

// ── expression evaluator (neutral irExpr AST) ───────────────────────────────

function evalExpr(x, scope) {
  switch (x.k) {
    case 'num': return x.n.indexOf('.') >= 0 ? parseFloat(x.n) : parseInt(x.n, 10);
    case 'str': return x.s || '';
    case 'bool': return !!x.b;
    case 'path': return evalPath(x, scope);
    case 'call': {
      const fn = evalPath(x.recv, scope);
      const args = (x.args || []).map((a) => evalExpr(a, scope));
      return typeof fn === 'function' ? fn.apply(null, args) : undefined;
    }
    case 'unary':
      return x.op === '!' ? !truthy(evalExpr(x.x, scope)) : -num(evalExpr(x.x, scope));
    case 'bin': return evalBin(x.op, evalExpr(x.l, scope), evalExpr(x.r, scope));
  }
  return undefined;
}

function evalPath(x, scope) {
  const segs = x.segs || [];
  let cur = x.local ? scope.locals[segs[0]] : scope.data[segs[0]];
  for (let i = 1; i < segs.length; i++) {
    if (cur == null) return undefined;
    cur = cur[segs[i]];
  }
  return cur;
}

function evalBin(op, a, b) {
  switch (op) {
    case '==': return a === b;
    case '!=': return a !== b;
    case '<': return num(a) < num(b);
    case '<=': return num(a) <= num(b);
    case '>': return num(a) > num(b);
    case '>=': return num(a) >= num(b);
    case '&&': return truthy(a) && truthy(b);
    case '||': return truthy(a) ? a : b;
    case '+': return (typeof a === 'string' || typeof b === 'string') ? '' + a + b : num(a) + num(b);
    case '-': return num(a) - num(b);
    case '*': return num(a) * num(b);
    case '/': return num(a) / num(b);
    case '%': return num(a) % num(b);
  }
  return undefined;
}

function num(v) { return typeof v === 'number' ? v : (v === true ? 1 : v === false ? 0 : parseFloat(v) || 0); }
function truthy(v) {
  if (Array.isArray(v)) return v.length > 0;
  if (typeof v === 'number') return v !== 0;
  if (typeof v === 'string') return v.length > 0;
  return !!v;
}

function htmlEscape(s) {
  return String(s)
    .replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;').replace(/'/g, '&#39;');
}

// ── renderer ────────────────────────────────────────────────────────────────

function renderNodes(nodes, scope, app) {
  let out = '';
  for (const n of nodes) {
    switch (n.t) {
      case 'text': out += n.v; break;
      case 'expr': out += htmlEscape(fmt(evalExpr(n.x, scope))); break;
      case 'if':
        out += truthy(evalExpr(n.x, scope)) ? renderNodes(n.then, scope, app) : renderNodes(n.els, scope, app);
        break;
      case 'for': {
        const list = evalExpr(n.x, scope) || [];
        for (const item of list) {
          const locals = Object.assign({}, scope.locals);
          locals[n.var] = item;
          out += renderNodes(n.body, { data: scope.data, locals: locals }, app);
        }
        break;
      }
      case 'child': {
        const childData = {};
        for (const p of n.props) childData[p.name] = p.x ? evalExpr(p.x, scope) : p.lit;
        out += app.renderFacet(n.name, childData).html;
        break;
      }
    }
  }
  return out;
}

function fmt(v) {
  if (v == null) return '';
  if (typeof v === 'boolean') return v ? 'true' : 'false';
  return '' + v;
}

// Resolve a facet_id pattern ("LikeButton:post:{post.id}") against data.
function resolveFacetID(pattern, data) {
  return pattern.replace(/\{([^}]+)\}/g, function (_, p) {
    const segs = p.split('.');
    let cur = data[segs[0]];
    for (let i = 1; i < segs.length && cur != null; i++) cur = cur[segs[i]];
    return fmt(cur);
  });
}

// Inject data-facet-id onto the first element of a rendered body (mirrors the Go
// path's injectFacetID).
function injectFacetID(html, id) {
  return html.replace(/<([a-zA-Z][\w-]*)/, function (m) { return m + ' data-facet-id="' + id + '"'; });
}

// ── signing (mirrors fa/event.go sign(): op \0 facet_id \0 fragment) ─────────

function signEvent(keyHex, ev) {
  if (!keyHex) return ev;
  const mac = crypto.createHmac('sha256', Buffer.from(keyHex, 'hex'));
  mac.update(ev.op); mac.update(Buffer.from([0]));
  mac.update(ev.facet_id); mac.update(Buffer.from([0]));
  mac.update(ev.fragment);
  ev.hmac = mac.digest('hex');
  return ev;
}

// ── App ──────────────────────────────────────────────────────────────────────

class App {
  // opts: { genDir, faKey (hex; '' = dev/no signing), runtimeJs (path to
  // fa-runtime.js), title }
  constructor(opts) {
    opts = opts || {};
    this.genDir = opts.genDir || 'generated';
    this.keyHex = opts.faKey != null ? opts.faKey : (process.env.FA_SIGNING_KEY || '');
    this.title = opts.title || 'FA';
    this.runtimeJs = opts.runtimeJs || path.join(__dirname, '..', '..', 'runtime', 'fa-runtime.js');

    this.manifest = JSON.parse(fs.readFileSync(path.join(this.genDir, 'manifest.json'), 'utf8'));
    this.ir = JSON.parse(fs.readFileSync(path.join(this.genDir, 'render.json'), 'utf8'));
    this.trees = {};   // facet name → block tree
    this.facets = {};  // facet name → ir facet
    for (const f of this.ir.facets) { this.facets[f.name] = f; this.trees[f.name] = buildTree(f.render); }

    this.handlers = {}; // event type → fn(ctx) → data
    this._root = null;  // { name, dataFn }
    this.conns = new Map(); // conn id → { res }
  }

  // root sets the facet rendered at GET / and its initial data.
  root(name, dataFn) { this._root = { name: name, dataFn: dataFn }; return this; }

  // on registers a when: handler. fn(ctx) returns the fresh facet data; the
  // runtime re-renders per the facet's when-mutations and pushes signed events
  // to the acting connection.
  on(type, fn) { this.handlers[type] = fn; return this; }

  // renderFacet interprets a facet's IR with data → { facet_id, html } with
  // data-facet-id injected on the root element.
  renderFacet(name, data) {
    const tree = this.trees[name];
    if (!tree) throw new Error('unknown facet ' + name);
    const html = injectFacetID(renderNodes(tree, { data: data, locals: {} }, this), resolveFacetID(this.facets[name].facet_id, data));
    return { facet_id: resolveFacetID(this.facets[name].facet_id, data), html: html };
  }

  // ── HTTP ────────────────────────────────────────────────────────────────
  listen(addr) {
    const a = addr || process.env.FA_ADDR || 'localhost:7373';
    const [host, port] = a.indexOf(':') >= 0 ? a.split(':') : ['localhost', a];
    const server = http.createServer(this._handle.bind(this));
    server.listen(parseInt(port, 10), host, () => console.log('fa(node): listening on http://' + host + ':' + port));
    return server;
  }

  _handle(req, res) {
    const url = req.url.split('?')[0];
    if (req.method === 'GET' && url === '/') return this._serveShell(req, res);
    if (req.method === 'GET' && url === '/sse') return this._serveSSE(req, res);
    if (req.method === 'POST' && url === '/events') return this._handleEvents(req, res);
    if (req.method === 'GET' && url === '/manifest.json') return this._json(res, this.manifest);
    if (req.method === 'GET' && url === '/render.json') return this._json(res, this.ir);
    if (req.method === 'GET' && url === '/fa-runtime.js') {
      res.writeHead(200, { 'Content-Type': 'text/javascript' });
      return res.end(fs.readFileSync(this.runtimeJs));
    }
    res.writeHead(404); res.end('not found');
  }

  _serveShell(req, res) {
    // A native client (FA-Native: 1) gets the neutral ViewNode tree as a
    // ScreenResponse {title, tree}; a browser gets the HTML shell.
    if (req.headers['fa-native'] === '1') {
      let tree = { kind: 'box' };
      if (this._root) {
        const html = this.renderFacet(this._root.name, this._root.dataFn({})).html;
        tree = native.nodeToJSON(native.parseView(html));
      }
      return this._json(res, { title: this.title, tree: tree });
    }
    let body = '';
    if (this._root) body = this.renderFacet(this._root.name, this._root.dataFn({})).html;
    const html = '<!doctype html><html><head><meta charset="utf-8">' +
      '<meta name="fa-key" content="' + this.keyHex + '">' +
      '<title>' + htmlEscape(this.title) + '</title></head><body>' +
      body + '<script src="/fa-runtime.js"></script></body></html>';
    res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
    res.end(html);
  }

  _serveSSE(req, res) {
    const v = (req.url.split('?')[1] || '').split('&').map((s) => s.split('=')).reduce((m, kv) => (m[kv[0]] = kv[1], m), {}).v || '1';
    if (v !== WIRE_VERSION) { res.writeHead(426); return res.end('fa: unsupported wire version ' + v); }
    res.writeHead(200, { 'Content-Type': 'text/event-stream', 'Cache-Control': 'no-cache', 'Connection': 'keep-alive', 'X-Accel-Buffering': 'no' });
    const id = crypto.randomBytes(16).toString('hex');
    this.conns.set(id, { res: res, native: req.headers['fa-native'] === '1' });
    const hello = JSON.stringify({ op: '_conn', conn: id, key: this.keyHex, v: WIRE_VERSION });
    res.write('data: ' + hello + '\n\n');
    const hb = setInterval(() => res.write(': keepalive\n\n'), 25000);
    req.on('close', () => { clearInterval(hb); this.conns.delete(id); });
  }

  _handleEvents(req, res) {
    let buf = '';
    req.on('data', (c) => { buf += c; });
    req.on('end', () => {
      let body;
      try { body = JSON.parse(buf || '{}'); } catch (e) { res.writeHead(400); return res.end('bad json'); }
      const fn = this.handlers[body.type];
      res.writeHead(204); res.end();
      if (!fn) return;
      const ctx = { type: body.type, payload: body.payload || {}, conn: body.conn };
      const data = fn(ctx);
      if (data == null) return;
      this._pushReRender(body.type, body.conn, data);
    });
  }

  // Map an incoming event type to the facet whose when: declares it, then apply
  // its mutations as signed events to the acting connection.
  _pushReRender(type, conn, data) {
    const c = this.conns.get(conn);
    if (!c) return;
    for (const f of this.ir.facets) {
      for (const w of f.when || []) {
        if (w.events.indexOf(type) < 0) continue;
        for (const mu of w.mutations) {
          const target = mu.target || f.name;
          const r = this.renderFacet(target, data);
          // A native connection receives the neutral ViewNode tree as the
          // fragment (and signs over it); a web connection receives HTML.
          const fragment = c.native ? native.treeJSON(r.html) : r.html;
          const ev = signEvent(this.keyHex, { op: mu.op === 'replace_all' ? 'replace' : mu.op, facet_id: r.facet_id, fragment: fragment });
          c.res.write('data: ' + JSON.stringify(ev) + '\n\n');
        }
      }
    }
  }

  _json(res, obj) { res.writeHead(200, { 'Content-Type': 'application/json' }); res.end(JSON.stringify(obj)); }
}

module.exports = { App, renderNodes, buildTree, evalExpr, signEvent, resolveFacetID, htmlEscape, WIRE_VERSION };
